torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -178,39 +198,62 @@ void RequestCallbackImpl::processRpc(
+              // self. In that case both the caller and the callee will access
+              // the OwnerRRef. Hence, on the callee side (here), it should not
+              // call addForkOfOwner as it is not a fork. To allow callee to
+              // distinguish when this request is sent to self, the caller will
+              // set forkId using rrefId (OwnerRRef does not have a forkId
+              // anyway).
+              RRefContext::getInstance().addForkOfOwner(rrefId, forkId);
+            }
+            Message m = RemoteRet(rrefId, forkId).toMessage();
+            m.setId(messageId);
+            responseFuture->markCompleted(std::move(m));
+          };
+
+      // ScriptRemoteCall is only alive within this block, use reference to
       // avoid copy
       auto& stack = scriptRemoteCall.stackRef();
       if (scriptRemoteCall.hasOp()) {
         scriptRemoteCall.op()->getOperation()(stack);
-      } else {
-        PythonRpcHandler::getInstance()
-            .jitCompilationUnit()
-            ->get_function(scriptRemoteCall.qualifiedName())
-            .run(stack);
+        TORCH_INTERNAL_ASSERT(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
+            stack.size());
+        scriptRemoteCallDone(std::move(stack.front()));
+        return;
       }
 
-      TORCH_INTERNAL_ASSERT(
-          stack.size() == 1,
-          "Return value of a builtin operator or a "
-          "TorchScript function should be a single IValue, got a vector of "
-          "size ",
-          stack.size());
-
-      ownerRRef->setValue(std::move(stack.front()));
-      if (rrefId != forkId) {
-        // Caller is a user and callee is the owner, add fork
-        //
-        // NB: rrefId == forkId is true if and only if calling remote to self.
-        // In that case both the caller and the callee will access the
-        // OwnerRRef. Hence, on the callee side (here), it should not call
-        // addForkOfOwner as it is not a fork. To allow callee to distinguish
-        // when this request is sent to self, the caller will set forkId using
-        // rrefId (OwnerRRef does not have a forkId anyway).
-        ctx.addForkOfOwner(rrefId, forkId);
+      auto jitFuture = PythonRpcHandler::getInstance()
+                           .jitCompilationUnit()
+                           ->get_function(scriptRemoteCall.qualifiedName())
+                           .runAsync(stack);
+      if (jitFuture->completed()) {
+        scriptRemoteCallDone(jitFuture->value());
+      } else {
+        jitFuture->addCallback(
+            [scriptRemoteCallDone, jitFuture, responseFuture]() {
+              try {
+                scriptRemoteCallDone(jitFuture->value());
+              } catch (const std::exception& e) {
+                // TODO: the better approach here would be to allow ownerRRef to
```

Comment:
Since this is added in #35331, do we need to update this accordingly? I am OK with either doing it here or in a followup PR

Reply:
I'm removing this comment, thanks.